src/Functions/FunctionsCharsetClassification.cpp

```diff
@@ -0,0 +1,247 @@
+            if (res < max_result)
+            {
+                return res;
+            }
+        }
+        return res;
+    }
+
+
+    static ALWAYS_INLINE size_t readCodePoints(CodePoint * code_points, const char *& pos, const char * end)
+    {
+        constexpr size_t padding_offset = default_padding - N + 1;
+        memcpy(code_points, code_points + padding_offset, roundUpToPowerOfTwoOrZero(N - 1) * sizeof(CodePoint));
+        memcpy(code_points + (N - 1), pos, default_padding * sizeof(CodePoint));
+        pos += padding_offset;
+        if (pos > end)
+            return default_padding - (pos - end);
+        return default_padding;
+    }
+
+    /// Сount how many times each bigram occurs in the text.
+    static ALWAYS_INLINE inline size_t calculateStats(
+        const char * data,
+        const size_t size,
+        size_t (*read_code_points)(CodePoint *, const char *&, const char *),
+        std::unordered_map<UInt16, Float64>& model)
+    {
+
+        const char * start = data;
+        const char * end = data + size;
+        CodePoint cp[simultaneously_codepoints_num] = {};
+        /// read_code_points returns the position of cp where it stopped reading codepoints.
+        size_t found = read_code_points(cp, start, end);
+        /// We need to start for the first time here, because first N - 1 codepoints mean nothing.
+        size_t i = N - 1;
+        size_t len = 0;
+        do
+        {
+            for (; i + N <= found; ++i)
+            {
+                UInt32 hash = 0;
+                for (size_t j = 0; j < N; ++j)
+                {
+                    hash <<= 8;
+                    hash += *(cp + i + j);
+                }
+                if (model[hash] == 0)
+                {
+                    model[hash] = 1;
+                    ++len;
+                }
+                ++model[hash];
+            }
+            i = 0;
+        } while (start < end && (found = read_code_points(cp, start, end)));
+
+        return len;
+    }
+
+
+    static void constant(String data, String & res)
+    {
+        const auto & encodings_freq = FrequencyHolder::getInstance().getEncodingsFrequency();
+
+        std::unordered_map<UInt16, Float64> model;
+        calculateStats(data.data(), data.size(), readCodePoints, model);
+
+        Float64 max_result = log(zero_frequency) * (max_string_size);
+        String poss_ans;
+        /// Go through the dictionary and find the charset with the highest weight
+        for (auto& item : encodings_freq)
+        {
+            Float64 score = naiveBayes(item.map, model, max_result);
+            if (max_result < score)
+            {
+                poss_ans = item.name;
+                max_result = score;
+            }
+        }
+
+        /* In our dictionary we have lines with form: <Language>_<Charset>
+         * If we need to find language of data, we return <Language>
+         * If we need to find charset of data, we return <Charset>.
+         */
```

Comment:
Thank you for this comment, it's good to know)
But maybe can we preprocess this beforehand?

We can build an array with lang strings, an array with charset strings, and store a pair of ints in the map.

Reply:
And also we can just return StringRef or string_view from this function.